README.md


# MJ Image Slider

![](https://github.com/miladjalalli/MJ-Image-Slider/blob/master/images/example.gif?raw=true)

## Supported platforms

* Flutter Android
* Flutter iOS
* Flutter web
* Flutter desktop

## Installation

Add `mj_image_slider:` to your `pubspec.yaml` dependencies. And import it:

```dart
import 'package:mj_image_slider/MJOptions.dart';
import 'package:mj_image_slider/mj_image_slider.dart';
```

## How to use

Simply create a `MJImageSLider` widget, and pass the required params:

```dart

List<String> images = [
    "assets/images/images_1.jpg",
    "assets/images/images_2.jpg",
    "assets/images/images_3.jpg",
    "assets/images/images_4.jpg",
  ];
  
MJImageSlider(
  options: MjOptions(height: 200.0,wifth:400),
  widgets: [...images.map((e) => Image(image: AssetImage(e))).toList()],
)
```



## Params

```dart

MJImageSlider(
   options: MjOptions(
      height: 200,
      width:400
      aspectRatio: 16/9,
      viewportFraction: 0.8,
      initialPage: 0,
      enableInfiniteScroll: true,
      autoPlayInterval: Duration(seconds: 3),
      autoPlayAnimationDuration: Duration(milliseconds: 800),
      autoPlayCurve: Curves.fastOutSlowIn,
      onPageChanged: callbackFunction,
      scrollDirection: Axis.horizontal,
   ),
   widgets:widgets,
 )
```

CHANGELOG.md
## 0.0.7

Add Options For Slider

test/mj_image_slider_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:mj_image_slider/mj_image_slider.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


lib/MJOptions.dart
import 'package:flutter/material.dart';


class MjOptions{
  /// Set slider height and overrides any existing [aspectRatio].
  double? height;

  /// Set slider width and overrides any existing [aspectRatio].
  double? width;

  /// Aspect ratio is used if no height have been declared.
  ///
  /// Defaults to 16:9 aspect ratio.
  late double aspectRatio;

  /// The fraction of the viewport that each page should occupy.
  ///
  /// Defaults to 0.8, which means each page fills 80% of the slider.
  late double viewportFraction;

  /// The initial page to show when first creating the [Slider].
  ///
  /// Defaults to 0.
  int initialPage = 0;

  ///Determines if slider should loop infinitely or be limited to item length.
  ///
  ///Defaults to true, i.e. infinite loop.
  late bool enableInfiniteScroll;

  /// Sets Duration to determent the frequency of slides when
  ///
  /// [autoPlay] is set to true.
  /// Defaults to 4 seconds.
  late Duration autoPlayInterval;

  /// The animation duration between two transitioning pages while in auto playback.
  ///
  /// Defaults to 800 ms.
  late Duration autoPlayAnimationDuration;

  /// Determines the animation curve physics.
  ///
  /// Defaults to [Curves.fastOutSlowIn].
  late Curve autoPlayCurve;


  /// The axis along which the page view scrolls.
  ///
  /// Defaults to [Axis.horizontal].
  late Axis scrollDirection;

  /// Called whenever the page in the center of the viewport changes.
  Function(int index)? onPageChanged;

}

lib/mj_options.dart
import 'package:flutter/material.dart';

class MjOptions {
  MjOptions({
    this.height,
    this.width,
    this.aspectRatio: 16 / 9,
    this.viewportFraction: 0.8,
    this.initialPage: 0,
    this.enableInfiniteScroll: true,
    this.autoPlayInterval: const Duration(seconds: 4),
    this.autoPlayAnimationDuration = const Duration(milliseconds: 800),
    this.autoPlayCurve: Curves.fastOutSlowIn,
    this.onPageChanged,
    this.scrollDirection: Axis.horizontal,
  });

  /// Set slider height and overrides any existing [aspectRatio].
  double? height;

  /// Set slider width and overrides any existing [aspectRatio].
  double? width;

  /// Aspect ratio is used if no height have been declared.
  ///
  /// Defaults to 16:9 aspect ratio.
  late double aspectRatio;

  /// The fraction of the viewport that each page should occupy.
  ///
  /// Defaults to 0.8, which means each page fills 80% of the slider.
  late double viewportFraction;

  /// The initial page to show when first creating the [Slider].
  ///
  /// Defaults to 0.
  int initialPage;

  ///Determines if slider should loop infinitely or be limited to item length.
  ///
  ///Defaults to true, i.e. infinite loop.
  late bool enableInfiniteScroll;

  /// Sets Duration to determent the frequency of slides when
  ///
  /// [autoPlay] is set to true.
  /// Defaults to 4 seconds.
  late Duration autoPlayInterval;

  /// The animation duration between two transitioning pages while in auto playback.
  ///
  /// Defaults to 800 ms.
  late Duration autoPlayAnimationDuration;

  /// Determines the animation curve physics.
  ///
  /// Defaults to [Curves.fastOutSlowIn].
  late Curve autoPlayCurve;

  /// The axis along which the page view scrolls.
  ///
  /// Defaults to [Axis.horizontal].
  late Axis scrollDirection;

  /// Called whenever the page in the center of the viewport changes.
  Function(int index)? onPageChanged;
}


lib/mj_image_slider.dart
library mj_image_slider;

import 'dart:async';

import 'package:flutter/material.dart';
import 'package:mj_image_slider/mj_options.dart';

class MJImageSlider extends StatefulWidget {
  late List<Widget> widgets;
  MjOptions options;

  MJImageSlider({required this.widgets,
    required this.options});

  @override
  State<MJImageSlider> createState() => _MJImageSliderState();
}

class _MJImageSliderState extends State<MJImageSlider> {
  late PageController controller;
  double currentPageValue = 0.0;
  late Timer timer;

  @override
  void initState() {
    controller = PageController(
        viewportFraction: widget.options.viewportFraction,
        initialPage: widget.options.initialPage
    );

    controller.addListener(() {
      setState(() {
        currentPageValue = controller.page!;
      });
    });

    timer = Timer.periodic(widget.options.autoPlayInterval, (timer) {
      late int nextPage;
      if (widget.options.enableInfiniteScroll){
        nextPage = 0;
      }else{
        nextPage = currentPageValue.toInt();
      }


      if (currentPageValue < widget.widgets.length - 1) nextPage = currentPageValue.toInt() + 1;

      if (widget.options.autoPlayCurve == null) {
        controller.jumpToPage(nextPage);
      } else {
        controller.animateToPage(nextPage, duration: widget.options.autoPlayAnimationDuration, curve: widget.options.autoPlayCurve);
      }
    });
    super.initState();
  }

  @override
  void dispose() {
    timer.cancel();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return widget.options.height != null
        ? SizedBox(
        height: widget.options.height,
        width: widget.options.width,
        child: PageView.builder(
          controller: controller,
          itemBuilder: (context, index) {
            return widget.widgets[index];
          },
          onPageChanged: (index){
            widget.options.onPageChanged!(index);
          },
          scrollDirection: widget.options.scrollDirection,
          itemCount: widget.widgets.length,
        ))
        : AspectRatio(
      aspectRatio: widget.options.aspectRatio,
      child: PageView.builder(
        controller: controller,
        itemBuilder: (context, index) {
          return widget.widgets[index];
        },
        onPageChanged: (index){
          widget.options.onPageChanged!(index);
        },
        scrollDirection: widget.options.scrollDirection,
        itemCount: widget.widgets.length,
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:mj_image_slider/MJOptions.dart';
import 'package:mj_image_slider/mj_image_slider.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  List<String> images = [
    "assets/images/images_1.jpg",
    "assets/images/images_2.jpg",
    "assets/images/images_3.jpg",
    "assets/images/images_4.jpg",
  ];
  List<String> networkImages = [
    "https://images.unsplash.com/reserve/bOvf94dPRxWu0u3QsPjF_tree.jpg?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1176&q=80",
    "https://images.unsplash.com/photo-1546587348-d12660c30c50?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1174&q=80",
    "https://images.unsplash.com/photo-1476514525535-07fb3b4ae5f1?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1170&q=80",
    "https://images.unsplash.com/photo-1525127752301-99b0b6379811?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1170&q=80",
  ];

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: SingleChildScrollView(
        child: Padding(
          padding: const EdgeInsets.all(16.0),
          child: Wrap(
            crossAxisAlignment: WrapCrossAlignment.center,
            children: [
              MJImageSlider(
                options: MjOptions(onPageChanged: (i) {}, scrollDirection: Axis.vertical, height: 200, width: 400, viewportFraction: 4.0),
                widgets: [...images.map((e) => Image(image: AssetImage(e))).toList()],
              ),
              const SizedBox(
                width: 32,
              ),
              MJImageSlider(
                options: MjOptions(
                    onPageChanged: (i) {}, enableInfiniteScroll: false, autoPlayCurve: Curves.easeInToLinear, height: 200, width: 400, viewportFraction: 1.0),
                widgets: [...networkImages.map((e) => Image(image: NetworkImage(e))).toList()],
              ),
              const SizedBox(
                width: 32,
              ),
              MJImageSlider(
                options: MjOptions(onPageChanged: (i) {}, autoPlayCurve: Curves.easeInExpo, height: 200, width: 400, viewportFraction: 0.8),
                widgets: [
                  ...networkImages
                      .map((e) => Image(image: NetworkImage(e)))
                      .toList()
                ],
              ),
            ],
          ),
        ),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
